Model/Observer/CategorySaveAfterObserver.php

```diff
@@ -0,0 +1,116 @@
+use Magento\Framework\Registry;
+use Wizzy\Search\Services\Indexer\IndexerManager;
+use Magento\Catalog\Model\Category;
+
+class CategorySaveAfterObserver implements ObserverInterface
+{
+    protected $categoryRepository;
+    protected $registry;
+    protected $indexer;
+
+    public function __construct(
+        CategoryRepositoryInterface $categoryRepository,
+        Registry $registry,
+        IndexerManager $indexerManager
+    ) {
+        $this->categoryRepository = $categoryRepository;
+        $this->registry = $registry;
+        $this->indexer = $indexerManager->getProductsIndexer();
+    }
+
+    public function execute(Observer $observer)
+    {
+        $category = $observer->getEvent()->getCategory();
+        $categoryId = $category->getId();
+
+        $oldData = $this->registry->registry('original_category_data');
+        $oldProductIds = $this->registry->registry('original_product_ids');
+
+        if (!$oldData) {
+            return;
+        }
+
+        $newData = $category->getData();
+        $syncAll = false;
+        $criticalFields = ['name', 'url_key', 'is_active'];
+
+        foreach ($criticalFields as $field) {
+            if (($oldData[$field] ?? null) != ($newData[$field] ?? null)) {
+                $syncAll = true;
+                break;
+            }
+        }
+
+        $productIdsToSync = [];
+
+        if ($syncAll) {
+            $productIdsToSync = $this->getAllProductIdsPaginated($category);
+        } else {
+            $originalCategory = $this->categoryRepository->get($categoryId, true);
+            $oldPositions = $originalCategory->getProductsPosition() ?? [];
+            $newPositions = $category->getProductsPosition() ?? [];
+
+            foreach ($newPositions as $productId => $newPosition) {
+                $oldPosition = $oldPositions[$productId] ?? null;
+                if ($oldPosition !== null && (int)$oldPosition !== (int)$newPosition) {
+                    $productIdsToSync[] = (int)$productId;
+                }
+            }
+
+            if (is_array($oldProductIds)) {
+                $newProductIds = $this->getAllProductIdsPaginated($category);
+                $diffProductIds = array_values(array_unique(array_merge(
+                    array_diff($newProductIds, $oldProductIds),
+                    array_diff($oldProductIds, $newProductIds)
+                )));
+
+                $productIdsToSync = array_values(array_unique(array_merge($productIdsToSync, $diffProductIds)));
+            }
+        }
+
+        if (!empty($productIdsToSync) && !$this->indexer->isScheduled()) {
+            $this->indexer->reindexList($productIdsToSync);
+        }
+
+        $this->registry->unregister('original_category_data');
+        $this->registry->unregister('original_product_ids');
+    }
+
+    /**
+     * Fetch all product IDs in paginated chunks
+     */
+    protected function getAllProductIdsPaginated(Category $category, int $pageSize = 5000): array
```

Comment:
If 

        $originalCategory->getProductCollection()->getAllIds();

is going to be DB Heavy. We will need to put this inside entity service file and use it in both the files.